### PR TITLE
Make gossipd more forgiving of temporary routing failures

### PR DESCRIPTION
**Pull request by ZmnSCPxj**
Fixes: #867 

Policy: disable failing channels for 20 seconds, or the next `channel_update`.

Implementation: Use a `u64 unroutable_until`, a UNIX-time variable, which prevents a channel from being used for routing if less than the current time in UNIX epoch.  Set this to 20+current time if routing failure reported.  Set to 0 on construction or `channel_update`.

### CODE REVIEW

**Comment by ZmnSCPxj** [comment]
Rebased to fix conflict.

File: gossipd/routing.c [hunk 1]
```diff
@@ -341,6 +343,12 @@ static void bfg_one_edge(struct node *node, size_t edgenum, double riskfactor)
 	}
 }
 
+/* Determine if the given node_connection is routable */
+static bool nc_is_routable(const struct node_connection *nc, u64 now)
```
Comment (rustyrussell) [inline gossipd/routing.c]: I would use time_t here; it's u64 where it matters and is self-documenting nicely.

File: gossipd/routing.c [hunk 2]
```diff
@@ -351,6 +359,10 @@ find_route(const tal_t *ctx, struct routing_state *rstate,
 	struct node_map_iter it;
 	struct node_connection *first_conn;
 	int runs, i, best;
+	/* Call time_now() once at the start, so that our tight loop
+	 * does not keep calling into operating system for the
+	 * current time */
+	u64 now = time_now().ts.tv_sec;
```
Comment (rustyrussell) [inline gossipd/routing.c]: Here too....

File: gossipd/routing.c [hunk 3]
```diff
@@ -1041,15 +1054,17 @@ get_out_node_connection_of(struct routing_state *rstate,
  */
 static void routing_failure_on_nc(struct routing_state *rstate,
 				  enum onion_type failcode,
-				  struct node_connection *nc)
+				  struct node_connection *nc,
+				  u64 now)
```
Comment (rustyrussell) [inline gossipd/routing.c]: Here too...

File: gossipd/routing.h [hunk 4]
```diff
@@ -38,6 +38,10 @@ struct node_connection {
 	/* Cached `channel_announcement` and `channel_update` we might forward to new peers*/
 	u8 *channel_announcement;
 	u8 *channel_update;
+
+	/* If greater than current time, this connection should not
+	 * be used for routing. */
+	u64 unroutable_until;
```
Comment (rustyrussell) [inline gossipd/routing.h]: Here too...

**Comment by ZmnSCPxj** [comment]
Fixed as per review.

**Comment by cdecker** [comment]
ACK d9e4203

**Pull request merged**